x-pack/plugins/task_manager/server/task_validator.ts

```diff
@@ -0,0 +1,183 @@
+          latestStateSchema,
+          'ignore'
+        );
+      } catch (e) {
+        if (!this.allowReadingInvalidState) {
+          throw e;
+        }
+        this.logger.debug(
+          `[${task.taskType}][${task.id}] Failed to validate the task's state. Allowing read operation to proceed because allow_reading_invalid_state is true. Error: ${e.message}`
+        );
+      }
+
+      return {
+        ...task,
+        state,
+      };
+    }
+
+    // We are doing a write operation which must validate against the latest state schema
+    return {
+      ...task,
+      state: this.getValidatedStateSchema(task.state, task.taskType, latestStateSchema, 'forbid'),
+      stateVersion: latestStateSchema?.version,
+    };
+  }
+
+  private migrateTaskState(
+    state: ConcreteTaskInstance['state'],
+    currentVersion: number | undefined,
+    taskTypeDef: TaskDefinition,
+    latestStateSchema: LatestStateSchema
+  ) {
+    if (!latestStateSchema || (currentVersion && currentVersion >= latestStateSchema.version)) {
+      return state;
+    }
+
+    let migratedState = state;
+    for (let i = currentVersion || 1; i <= latestStateSchema.version; i++) {
+      if (!taskTypeDef.stateSchemaByVersion || !taskTypeDef.stateSchemaByVersion[`${i}`]) {
+        throw new Error(
+          `[TaskValidator] state schema for ${taskTypeDef.type} missing version: ${i}`
+        );
+      }
+      migratedState = taskTypeDef.stateSchemaByVersion[i].up(migratedState);
+      try {
+        taskTypeDef.stateSchemaByVersion[i].schema.validate(migratedState);
+      } catch (e) {
+        throw new Error(
+          `[TaskValidator] failed to migrate to version ${i} because the data returned from the up migration doesn't match the schema: ${e.message}`
+        );
+      }
+    }
+
+    return migratedState;
+  }
+
+  private getValidatedStateSchema(
+    state: ConcreteTaskInstance['state'],
+    taskType: string,
+    latestStateSchema: LatestStateSchema,
+    unknowns: 'forbid' | 'ignore'
+  ): ConcreteTaskInstance['state'] {
+    if (!latestStateSchema) {
+      throw new Error(
+        `[TaskValidator] stateSchemaByVersion not defined for task type: ${taskType}`
+      );
+    }
+
+    return this.cachedExtendSchema({ unknowns, taskType, latestStateSchema }).validate(state);
+  }
+}
+
+function extendSchema(options: {
+  latestStateSchema: LatestStateSchema;
+  unknowns: 'forbid' | 'ignore';
+  taskType: string;
+}) {
+  if (!options.latestStateSchema) {
+    throw new Error(
+      `[TaskValidator] stateSchemaByVersion not defined for task type: ${options.taskType}`
+    );
+  }
+  return options.latestStateSchema.schema.extendsDeep({ unknowns: options.unknowns });
+}
+
+function getLatestStateSchema(taskTypeDef: TaskDefinition): LatestStateSchema {
+  if (!taskTypeDef.stateSchemaByVersion) {
+    return;
+  }
+
+  const versions = Object.keys(taskTypeDef.stateSchemaByVersion).map((v) => parseInt(v, 10));
+  const latest = max(versions);
```

Comment:
if we use `latest` in the schema we don't need this :)

Reply:
I will still have to do some tricks to calculate the latest number to store in the `stateVersion` attribute of the SO. So if there's `{ latest: ..., 1: ... }` I'll have to calculate 2 as the version to store.